### PR TITLE
Add Notebook: Google NMT on IWSLT 2015 English-Vietnamese Translation

### PR DESCRIPTION
**Pull request by szhengac**
## Description ##
Add notebook: Google NMT on IWSLT 2015 English-Vietnamese Translation. @cgraywang will help polish the notebook.

### Essentials ###
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by cgraywang** [comment]
@szhengac thanks!

File: docs/examples/nmt/gnmt.md [hunk 1]
```diff
@@ -0,0 +1,438 @@
...
+ctx = mx.gpu(0)
+
+# parameters for dataset
+dataset = 'IWSLT2015'
+src_lang = 'en'
+tgt_lang = 'vi'
+src_max_len = 50
+tgt_max_len = 50
+
+# parameters for model
```
Comment (eric-haibin-lin) [inline docs/examples/nmt/gnmt.md]: not related to this PR:
@szha maybe for gluonnlp we also want to define hyper_param registration like t2t for better abstraction? https://github.com/tensorflow/tensor2tensor/blob/master/tensor2tensor/models/transformer.py#L1467

File: docs/examples/nmt/gnmt.md [hunk 2]
```diff
@@ -0,0 +1,438 @@
...
+
+logging_config(save_dir)
+```
+
+## Load and Preprocess Dataset
+
+The following shows how to process the dataset and cache the processed dataset
+for the future use. The processing steps include: 1) clip the source and target
+sequences and 2) split the string input to a list of tokens and 3) map the
+string token into its index in the vocabulary and 4) append EOS token to source
```
Comment (eric-haibin-lin) [inline docs/examples/nmt/gnmt.md]: Since the notebook is for new comers, explaining abbr would help?
EOS token -> end-of-sentence (EOS) token.

File: docs/examples/nmt/gnmt.md [hunk 3]
```diff
@@ -0,0 +1,438 @@
...
+        return src_npy, tgt_npy
+
+
+def process_dataset(dataset, src_vocab, tgt_vocab, src_max_len=-1, tgt_max_len=-1):
+    start = time.time()
+    dataset_processed = dataset.transform(TrainValDataTransform(src_vocab, tgt_vocab,
+                                                                src_max_len,
+                                                                tgt_max_len), lazy=False)
+    end = time.time()
+    print('Processing Time spent: {}'.format(end - start))
```
Comment (eric-haibin-lin) [inline docs/examples/nmt/gnmt.md]: nit: either change `Time` to `time`, or capitalize `spent`?

File: docs/examples/nmt/gnmt.md [hunk 4]
```diff
@@ -0,0 +1,438 @@
...
+
+    Parameters
+    ----------
+    dataset : str
+    src_lang : str, default 'en'
+    tgt_lang : str, default 'vi'
+
+    Returns
+    -------
+
```
Comment (eric-haibin-lin) [inline docs/examples/nmt/gnmt.md]: returns ..?

File: docs/examples/nmt/gnmt.md [hunk 5]
```diff
@@ -0,0 +1,438 @@
...
+test_batchify_fn = btf.Tuple(btf.Pad(), btf.Pad(),
+                             btf.Stack(dtype='float32'), btf.Stack(dtype='float32'),
+                             btf.Stack())
+```
+
+We can then construct bucketing samplers, which generate batches by grouping
+sequences with similar lengths.
+
+```python
+train_batch_sampler = FixedBucketSampler(lengths=data_train_lengths,
```
Comment (eric-haibin-lin) [inline docs/examples/nmt/gnmt.md]: Some explanation of where 1.2 comes from would be great

File: docs/examples/nmt/gnmt.md [hunk 6]
```diff
@@ -0,0 +1,438 @@
...
+                                            num_layers=num_layers,
+                                            num_bi_layers=num_bi_layers)
+model = NMTModel(src_vocab=src_vocab, tgt_vocab=tgt_vocab, encoder=encoder, decoder=decoder,
+                 embed_size=num_hidden, prefix='gnmt_')
+model.initialize(init=mx.init.Uniform(0.1), ctx=ctx)
+static_alloc = True
+model.hybridize(static_alloc=static_alloc)
+logging.info(model)
+
+loss_function = SoftmaxCEMaskedLoss()
```
Comment (eric-haibin-lin) [inline docs/examples/nmt/gnmt.md]: Maybe some simple explanation of why masked loss is needed would be helpful

File: docs/examples/nmt/gnmt.md [hunk 7]
```diff
@@ -0,0 +1,438 @@
...
+
+```python
+trainer = gluon.Trainer(model.collect_params(), 'adam', {'learning_rate': lr})
+```
+
+We can then write the training iteration. During the training, we perform the
+one evaluation on validation and testing dataset every epoch, and record the
+parameters that give the hightest BLEU score on validation dataset. Before
+performing forward and backward, we first use `as_in_context` function to copy
+the mini-batch to GPU. The statement `with mx.autograd.record()` tell Gluon
```
Comment (eric-haibin-lin) [inline docs/examples/nmt/gnmt.md]: tell -> tells

File: docs/examples/nmt/gnmt.md [hunk 8]
```diff
@@ -0,0 +1,438 @@
...
+    if valid_bleu_score > best_valid_bleu:
+        best_valid_bleu = valid_bleu_score
+        save_path = os.path.join(save_dir, 'valid_best.params')
+        logging.info('Save best parameters to {}'.format(save_path))
+        model.save_parameters(save_path)
+    if epoch_id + 1 >= (epochs * 2) // 3:
+        new_lr = trainer.learning_rate * lr_update_factor
+        logging.info('Learning rate change to {}'.format(new_lr))
+        trainer.set_learning_rate(new_lr)
+```
```
Comment (eric-haibin-lin) [inline docs/examples/nmt/gnmt.md]: A summary/conclusion section will be great

File: docs/examples/nmt/gnmt.md (hunk 2 above)
Comment (szhengac) [reply docs/examples/nmt/gnmt.md]: Sure

File: docs/examples/nmt/gnmt.md (hunk 5 above)
Comment (szhengac) [reply docs/examples/nmt/gnmt.md]: this is empirically tuned.

File: docs/examples/nmt/gnmt.md (hunk 4 above)
Comment (szhengac) [reply docs/examples/nmt/gnmt.md]: We use capital one in gluonnlp.

File: docs/examples/nmt/gnmt.md (hunk 8 above)
Comment (szhengac) [reply docs/examples/nmt/gnmt.md]: Good point

File: docs/examples/nmt/gnmt.md (hunk 4 above)
Comment (eric-haibin-lin) [reply docs/examples/nmt/gnmt.md]: I meant the return types are missing

**Comment by mli** [comment]
Job PR-261/14 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-261/14/index.html

File: docs/examples/nmt/gnmt.md [hunk 9]
```diff
@@ -0,0 +1,444 @@
+# Google NMT on IWSLT 2015 English-Vietnamese Translation
+
+In this notebook, we going to train Google NMT on IWSLT 2015 English-Vietnamese
```
Comment (sxjscience) [inline docs/examples/nmt/gnmt.md]: we are going to train

**Comment by mli** [comment]
Job PR-261/16 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-261/16/index.html

**Comment by mli** [comment]
Job PR-261/17 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-261/17/index.html

File: docs/examples/nmt/gnmt.md (hunk 4 above)
Comment (szha) [reply docs/examples/nmt/gnmt.md]: @szhengac @sxjscience seems like the return types are still missing.

**Comment by szha** [comment]
cc'd @leezu @eric-haibin-lin who might have some suggestions on how to make a translation tutorial after KDD.

**Comment by mli** [comment]
Job PR-261/18 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-261/18/index.html

File: docs/examples/nmt/gnmt.md [hunk 10]
```diff
@@ -0,0 +1,456 @@
...
+import argparse
+import time
+import random
+import os
+import io
+import logging
+import numpy as np
+import mxnet as mx
+from mxnet import gluon
+from mxnet.gluon.data import ArrayDataset, SimpleDataset
```
Comment (szha) [inline docs/examples/nmt/gnmt.md]: import until the last modules, instead of importing the classes and methods.

File: docs/examples/nmt/gnmt.md [hunk 11]
```diff
@@ -0,0 +1,456 @@
...
+random.seed(100)
+mx.random.seed(10000)
+ctx = mx.gpu(0)
+
+# parameters for dataset
+dataset = 'IWSLT2015'
+src_lang = 'en'
+tgt_lang = 'vi'
+src_max_len = 50
+tgt_max_len = 50
```
Comment (szha) [inline docs/examples/nmt/gnmt.md]: for related hyperparameters, we can do them on the same line, such as `src_max_len, tgt_max_len = 50, 50`

**Comment by mli** [comment]
Job PR-261/20 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-261/20/index.html

**Comment by eric-haibin-lin** [comment]
For KDD we didnt have time to cover gmnt unfortunately

**Comment by mli** [comment]
Job PR-261/21 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-261/21/index.html

**Comment by mli** [comment]
Job PR-261/22 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-261/22/index.html

**Comment by mli** [comment]
Job PR-261/24 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-261/24/index.html

**Comment by mli** [comment]
Job PR-261/25 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-261/25/index.html

**Comment by mli** [comment]
Job PR-261/26 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-261/26/index.html
